### DISCUSSION TITLE
Userdata: Dynamically set values instead of static in the behave.ini

### DISCUSSION THREAD

**Discussion by przemolb** [behave/behave#963] [Q&A] [answered]
We are runing our behave test in a test environment running on k8s. It is static environment (pre-built) but we would like to dynamically create new environments on the fly which means we need to also dynamically create the content of behave.ini file.
We can have some static content in the file (which is shared across different environments). But some of the content needs to be dynamically generated. Is it possible to add or overwrite existing setting from behave.ini file using CLi parameters ? If yes does that apply to all parameters or some subset ?

**Comment by bittner** [comment]
Can you share which configuration specifically this would be?

Can you use environment variables and make the dynamic content of your setup dependent on that? Tox can handle environment variables and external `tox.ini` file, for example, which allows for arbitrary complexity.

**Reply by przemolb** [reply]
I am talking about `[behave.userdata]` i.e.:
```
[behave.userdata]
myvar1 = value1
myvar2 = value2
myvar3 = value3
```

**Comment by jenisys** [comment, marked as answer]
How you provide the userdata is up to you.

The use case that the userdata is provided in the "behave.ini" config-file is just the simplest case (that covers 50-70% of the cases, IMHO). For example, you can provide the userdata in the `before_all()` hook in "environment.py" (example: read it from another config-file).

EXAMPLE:

```sh
$ behave -D userdata_config_file=example1.ini features/
```

NOTE:
The `before_all()` hook that would make that work is left as exercise (but relevant parts are shown below).

SEE: [features/userdata.feature](https://github.com/behave/behave/blob/master/features/userdata.feature)

OTHERWISE:
* You can provide userdata dynamically and there is no need to change the „behave.ini“ config-file.
* The dynamic userdata can be provided from another data source (an other config-file that is selected)
   or generated by Python code, etc.
* How the dynamic userdata is injected into the behave userdata is up to you. 
  You need to provide some glue code that implements this mechanism and only you knows exactly 
  what and how you want to achieve. The glue code is normally needed in the `before_all()` hook.
* Cmdline option(s) for userdata override settings in the `[behave.userdata]` section of the „behave.ini“ config-file.
* `behave` does not provide any support for changing the config-file (reason: config-file is a read-only concept in the context of `behave` and not a persistent storage, changing the config-file can easily be provided outside of `behave`)
